examples/gbnf-validator/gbnf-validator.cpp

```diff
@@ -0,0 +1,122 @@
+#define LLAMA_API_INTERNAL
+
+#include "grammar-parser.h"
+#include "ggml.h"
+#include "llama.h"
+#include "unicode.h"
+
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <vector>
+
+static bool llama_sample_grammar_string(struct llama_grammar * grammar, const std::string & input_str, size_t & error_pos, std::string & error_msg) {
+    auto decoded = decode_utf8(input_str, {});
+    const auto & code_points = decoded.first;
+
+    size_t pos = 0;
+    for (auto it = code_points.begin(), end = code_points.end() - 1; it != end; ++it) {
+        auto prev_stacks = grammar->stacks;
+        grammar->stacks = llama_grammar_accept(grammar->rules, grammar->stacks, *it);
+        if (grammar->stacks.empty()) {
+            error_pos = pos;
+            error_msg = "Unexpected character '" + unicode_cpt_to_utf8(*it) + "'";
+            grammar->stacks = prev_stacks;
+            return false;
+        }
+        ++pos;
+    }
+
+    for (const auto & stack : grammar->stacks) {
+        if (stack.empty()) {
+            return true;
+        }
+    }
+
+    error_pos = pos;
+    error_msg = "Unexpected end of input";
+    return false;
+}
+
+static void print_error_message(const std::string & input_str, size_t error_pos, const std::string & error_msg) {
+    std::cout << "Input string is invalid according to the grammar." << std::endl;
+    std::cout << "Error: " << error_msg << " at position " << std::to_string(error_pos) << std::endl;
+    std::cout << std::endl;
+    std::cout << "Input string:" << std::endl;
+    std::cout << input_str.substr(0, error_pos);
+    if (error_pos < input_str.size()) {
+        std::cout << "\033[1;31m" << input_str[error_pos];
+        if (error_pos+1 < input_str.size()) {
+            std::cout << "\033[0;31m" << input_str.substr(error_pos+1);
```

Comment:
For consistency with the rest of the codebase, always use `fprintf` instead of output streams

Reply:
Done. I'm not the most familiar with mixing std::string with `fread` and whatnot, so hopefully I didn't do things too badly. Happy to revise again if there is a better way to do this (I'm sure there is).